python/dolfinx/fem/forms.py

```diff
@@ -333,3 +360,139 @@ def unique_spaces(V):
             return list(unique_spaces(V.transpose()))
 
     raise RuntimeError("Unsupported array of forms")
+
+
+@dataclass
+class CompiledForm:
+    """
+    Compiled UFL form without associated DOLFINx data
+
+    """
+
+    ufl_form: ufl.Form  # The original ufl form
+    ufcx_form: typing.Any  # The compiled form
+    module: typing.Any  #  The module
+    code: str  # The source code
+    dtype: npt.DTypeLike  # data type used for the `ufcx_form`
+
+
+def compile_form(
+    comm: MPI.Intracomm,
+    form: ufl.Form,
+    form_compiler_options: typing.Optional[dict] = None,
+    jit_options: typing.Optional[dict] = None,
+) -> CompiledForm:
+    """
+    Compile UFL form without associated DOLFINx data
+    """
+    p_ffcx = ffcx.get_options(form_compiler_options)
+    p_jit = jit.get_options(jit_options)
+    ufcx_form, module, code = jit.ffcx_jit(comm, form, p_ffcx, p_jit)
+    return CompiledForm(form, ufcx_form, module, code, p_ffcx["scalar_type"])
+
+
+def form_cpp_creator(
+    dtype: npt.DTypeLike,
+) -> typing.Union[
+    _cpp.fem.Form_float32,
+    _cpp.fem.Form_float64,
+    _cpp.fem.Form_complex64,
+    _cpp.fem.Form_complex128,
+]:
+    """Return the wrapped C++ class of a variational form of a specific scalar type.
```

Comment:
class as in OO sense or a set/category? 
It seems to me that it's returning a creator/construtor function (as per usage below).